{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cfceb4f6-d758-4cc4-b6fa-8bfc8516427a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def convert_amino_acid(sequence, to_three_letter=False):\n",
    "    \"\"\"\n",
    "    Convert between single-letter and three-letter amino acid abbreviations.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): The amino acid sequence to be converted.\n",
    "        to_three_letter (bool, optional): If True, convert from single-letter to three-letter abbreviations.\n",
    "                                          If False, convert from three-letter to single-letter abbreviations.\n",
    "\n",
    "    Returns:\n",
    "        str: The converted amino acid sequence.\n",
    "    \"\"\"\n",
    "\n",
    "    # Dictionary mapping single-letter to three-letter amino acid abbreviations\n",
    "    single_to_three = {\n",
    "        'A': 'Ala', 'R': 'Arg', 'N': 'Asn', 'D': 'Asp', 'C': 'Cys',\n",
    "        'E': 'Glu', 'Q': 'Gln', 'G': 'Gly', 'H': 'His', 'I': 'Ile',\n",
    "        'L': 'Leu', 'K': 'Lys', 'M': 'Met', 'F': 'Phe', 'P': 'Pro',\n",
    "        'S': 'Ser', 'T': 'Thr', 'W': 'Trp', 'Y': 'Tyr', 'V': 'Val'\n",
    "    }\n",
    "\n",
    "    # Dictionary mapping three-letter to single-letter amino acid abbreviations\n",
    "    three_to_single = {v: k for k, v in single_to_three.items()}\n",
    "\n",
    "    if to_three_letter:\n",
    "        return ''.join(single_to_three.get(aa, aa) for aa in sequence)\n",
    "    else:\n",
    "        converted = []\n",
    "        for i in range(0, len(sequence), 3):\n",
    "            three_letter = sequence[i:i+3]\n",
    "            single_letter = three_to_single.get(three_letter, three_letter)\n",
    "            converted.append(single_letter)\n",
    "        return ''.join(converted)\n",
    "\n",
    "def convert_hgvs_to_non_standard(hgvs):\n",
    "    # 提取变异类型和位置\n",
    "    match = re.match(r'p\\.([A-Z][a-z]{2})(\\d+)([A-Z][a-z]{2})', hgvs)\n",
    "    \n",
    "    if match:\n",
    "        # 解析提取的信息\n",
    "        ref_aa = match.group(1)\n",
    "        position = match.group(2)\n",
    "        alt_aa = match.group(3)\n",
    "        \n",
    "        # 将三字母代码转换为单字母代码\n",
    "        ref_aa_single = convert_amino_acid(ref_aa, to_three_letter=False)\n",
    "        alt_aa_single = convert_amino_acid(alt_aa, to_three_letter=False)\n",
    "        \n",
    "        # 生成非标准命名\n",
    "        non_standard = f\"p.{ref_aa_single}{position}{alt_aa_single}\"\n",
    "        return non_standard\n",
    "    else:\n",
    "        # 处理其它类型的变异描述\n",
    "        ref_aa = re.findall(r'([A-Z][a-z]{2})', hgvs)\n",
    "        ref_aa_single = [convert_amino_acid(aa, to_three_letter=False) for aa in ref_aa]\n",
    "        non_standard = hgvs\n",
    "        for aa, aa_single in zip(ref_aa, ref_aa_single):\n",
    "            non_standard = non_standard.replace(aa, aa_single)\n",
    "        return non_standard\n",
    "    \n",
    "\n",
    "def get_frameshift_seq(sequence, position):\n",
    "    \"\"\"\n",
    "    Get the frameshift sequence from a given position until the first '*' character.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): The input amino acid sequence.\n",
    "        position (int): The starting position to search from.\n",
    "\n",
    "    Returns:\n",
    "        str: The subsequence from the given position until the first '*' character.\n",
    "             If the position is out of range, an empty string is returned.\n",
    "    \"\"\"\n",
    "    if position < 0 or position >= len(sequence):\n",
    "        return \"\"\n",
    "\n",
    "    for i in range(position, len(sequence)):\n",
    "        if sequence[i] == '*':\n",
    "            return sequence[position:i]\n",
    "\n",
    "    return sequence[position:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "18831d1c-b2d7-4bff-81e9-2886ed777803",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Standard: p.Thr882_Glu906dup -> Non-Standard: p.T882_E906dup\n",
      "Standard: p.Gln94_Gln95del -> Non-Standard: p.Q94_Q95del\n",
      "Standard: p.Gly51fs -> Non-Standard: p.G51fs\n",
      "Standard: p.Glu40_Asn51del -> Non-Standard: p.E40_N51del\n"
     ]
    }
   ],
   "source": [
    "convert_amino_acid(\"Glu\")\n",
    "# 示例用法\n",
    "standard_hgvs_list = [\n",
    "    \"p.Thr882_Glu906dup\",\n",
    "    \"p.Gln94_Gln95del\",\n",
    "    \"p.Gly51fs\",\n",
    "    \"p.Glu40_Asn51del\"\n",
    "]\n",
    "\n",
    "non_standard_hgvs_list = [convert_hgvs_to_non_standard(hgvs) for hgvs in standard_hgvs_list]\n",
    "for standard, non_standard in zip(standard_hgvs_list, non_standard_hgvs_list):\n",
    "    print(f\"Standard: {standard} -> Non-Standard: {non_standard}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87603ac8-15e7-43ad-8874-36cce71ffdf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from Bio import SeqIO\n",
    "\n",
    "class MutationPeptideGenerator:\n",
    "    def __init__(self, csv_file, fasta_file, out_path, prefix, flank_length=15, encoding='utf-8'):\n",
    "        self.csv_file = csv_file\n",
    "        self.fasta_file = fasta_file\n",
    "        self.out_path = out_path\n",
    "        self.prefix = prefix\n",
    "        self.flank_length = flank_length\n",
    "        self.encoding = encoding\n",
    "        self.mutations = []\n",
    "        self.fasta_seqs = {}\n",
    "\n",
    "\n",
    "    def read_csv(self):\n",
    "        with open(self.csv_file, 'r', encoding=self.encoding) as csvfile:\n",
    "            reader = csv.DictReader(csvfile)\n",
    "            for row in reader:\n",
    "                if  row['CHECK'] == 'Y':\n",
    "                    self.mutations.append({\n",
    "                        'transcript_id': row['Feature_ID'],\n",
    "                        'hgvs_p': row['HGVS.p'],\n",
    "                        'variant_type': row['Annotation'],\n",
    "                        'gene_name': row['Gene_Name']\n",
    "                    })\n",
    "\n",
    "    def read_fasta(self):\n",
    "        ref_seq = None\n",
    "        for record in SeqIO.parse(self.fasta_file, 'fasta'):\n",
    "            if 'Ref' == record.description.split()[1]:\n",
    "                ref_seq = str(record.seq)\n",
    "                \n",
    "            else:\n",
    "                stranscript_id = str(record.id).split()[0]\n",
    "                hgvs_p = record.description.split(\"HGVS.p:\")[1]\n",
    "                variant_id = \"_\".join([stranscript_id,hgvs_p])\n",
    "                self.fasta_seqs[variant_id] = {\n",
    "                    'ref': ref_seq,\n",
    "                    'variant': str(record.seq)\n",
    "                }\n",
    "                ref_seq = None\n",
    "                \n",
    "    def dedup_var_peptides(self, ref_peptides, var_peptides):\n",
    "        ref_peptides_dedup = []\n",
    "        var_peptides_dedup = []\n",
    "        var_pep_uniq = []\n",
    "        for ref_peptide, var_peptide in zip(ref_peptides, var_peptides):\n",
    "            var_pep = var_peptide.split(\"\\n\")[1]\n",
    "            if var_pep not in  var_pep_uniq:\n",
    "                ref_peptides_dedup.append(ref_peptide)\n",
    "                var_peptides_dedup.append(var_peptide)\n",
    "                var_pep_uniq.append(var_pep)\n",
    "        return ref_peptides_dedup, var_peptides_dedup\n",
    "\n",
    "        \n",
    "    def get_var_peptides(self):\n",
    "        ref_peptides = []\n",
    "        var_peptides = []\n",
    "        for mutation in self.mutations:\n",
    "            transcript_id = mutation['transcript_id']\n",
    "            hgvs_p = mutation['hgvs_p']\n",
    "            gene_name = mutation['gene_name']\n",
    "            variant_id =  \"_\".join([transcript_id, hgvs_p])\n",
    "            variant_type = mutation['variant_type'].split(\"&\")\n",
    "\n",
    "            try:\n",
    "                ref_seq = self.fasta_seqs[variant_id]['ref']\n",
    "                var_seq = self.fasta_seqs[variant_id]['variant']\n",
    "            except:\n",
    "                print(f'{variant_id} not in fasta_seqs')\n",
    "                \n",
    "            ref_peptide,var_peptide = \"\",\"\"\n",
    "            \n",
    "            if len(variant_type) == 1:\n",
    "                var_type = variant_type[0]\n",
    "                if var_type == \"missense_variant\": ## eg: p.Gly13Asp\n",
    "                    \n",
    "                    ref_aa = hgvs_p[2:5]\n",
    "                    alt_aa = hgvs_p[-3:]\n",
    "                    mut_pos = int(hgvs_p[5:].rstrip(alt_aa))\n",
    "                    left_pos = mut_pos - self.flank_length -1\n",
    "                    left_pos = left_pos if left_pos > 0 else 0\n",
    "                    right_pos = mut_pos + self.flank_length\n",
    "                    \n",
    "                    ref_peptide = ref_seq[left_pos : right_pos]\n",
    "                    var_peptide = var_seq[left_pos: mut_pos-1] + convert_amino_acid(alt_aa) + var_seq[mut_pos : right_pos]\n",
    "                    \n",
    "                elif var_type == \"frameshift_variant\":  ## eg: p.Ser878fs\n",
    "                    ref_aa = hgvs_p[2:5]\n",
    "                    mut_pos = int(hgvs_p[5:].rstrip(\"fs\"))\n",
    "                    left_pos = mut_pos - self.flank_length -1\n",
    "                    left_pos = left_pos if left_pos > 0 else 0\n",
    "                    right_pos = mut_pos + self.flank_length\n",
    "                    ref_peptide = ref_seq[left_pos  : right_pos]\n",
    "                    var_peptide = get_frameshift_seq(var_seq, left_pos)\n",
    "                    \n",
    "                elif var_type == \"disruptive_inframe_insertion\":  ## eg: p.Thr882_Glu906dup\n",
    "                    left_aa = hgvs_p[2:5]\n",
    "                    right_aa = hgvs_p.split(\"_\")[1][:3]\n",
    "                    left_pos = int(hgvs_p[5:].split(\"_\")[0]) -1 \n",
    "                    left_pos = left_pos if left_pos_a > 0 else 0\n",
    "                    right_pos = int(hgvs_p.split(\"_\")[1].rstrip(\"dup\")[2:])\n",
    "                    dup_peptide = var_seq[left_pos: right_pos+1]\n",
    "                    if len(dup_peptide) >= self.flank_length:\n",
    "                        var_peptide = dup_peptide[-15:]*2\n",
    "                        ref_peptide = ref_seq[left_pos:(right_pos + self.flank_length)]\n",
    "                    else:\n",
    "                        extend_length = self.flank_length - len(dup_peptide)\n",
    "                        left_pos_a = left_pos - extend_length\n",
    "                        left_pos_a = left_pos_a if left_pos_a > 0 else 0\n",
    "                        \n",
    "                        var_peptide = var_seq[( left_pos-extend_length if (left_pos-extend_length) >0 else None):(right_pos+ len(dup_peptide)+extend_length)]\n",
    "                        ref_peptide = ref_seq[left_pos : (right_pos+ self.flank_length)]\n",
    "                        \n",
    "                elif var_type == \"conservative_inframe_deletion\": ## eg1: p.Gln94_Gln95del eg2: p.Asp31del\n",
    "                    if \"_\" in hgvs_p:\n",
    "                        print(hgvs_p,\"yesdf\")\n",
    "                        left_pos = int(hgvs_p[5:].split(\"_\")[0]) - 1\n",
    "                        right_pos = int(hgvs_p.split(\"_\")[1].rstrip(\"del\")[2:])\n",
    "                        del_base_num = right_pos - left_pos + 1\n",
    "                    else:\n",
    "                        left_pos = int(hgvs_p[5:].rstrip(\"del\")) -1 \n",
    "                        right_pos = int(hgvs_p[5:].rstrip(\"del\"))\n",
    "                        del_base_num = 1\n",
    "                        \n",
    "                    left_pos_shift = left_pos - self.flank_length\n",
    "                    left_pos_shift = left_pos_shift if left_pos_shift > 0 else 0\n",
    "                    right_pos_shift = right_pos + self.flank_length\n",
    "                    ref_peptide = ref_seq[left_pos_shift : right_pos_shift]\n",
    "                    var_peptide = var_seq[left_pos_shift : (right_pos_shift - del_base_num)]\n",
    "                elif var_type == \"synonymous_variant\":\n",
    "                    pass\n",
    "                elif var_type == \"stop_gained\":\n",
    "                    pass\n",
    "                    \n",
    "            elif  len(variant_type) > 1:\n",
    "                if \"frameshift_variant\" in variant_type:  ## eg: p.Gly51fs\n",
    "                    ref_aa = hgvs_p[2:5]\n",
    "                    mut_pos = int(hgvs_p[5:].rstrip(\"fs\"))\n",
    "                    left_pos = mut_pos - self.flank_length -1\n",
    "                    left_pos = left_pos if left_pos > 0 else 0\n",
    "                    right_pos = mut_pos + self.flank_length\n",
    "                    ref_peptide = ref_seq[left_pos  : right_pos]\n",
    "                    var_peptide = get_frameshift_seq(var_seq, left_pos)\n",
    "                    \n",
    "                elif \"conservative_inframe_deletion\" in variant_type: ## eg: p.Glu40_Asn51del\n",
    "                    \"\"\"\n",
    "                    annotation: splice_acceptor_variant&splice_donor_variant&conservative_inframe_deletion&splice_region_variant&intron_variant\n",
    "                    eg: p.Glu40_Asn51del\n",
    "                    \"\"\"\n",
    "                    left_pos = int(hgvs_p[5:].split(\"_\")[0])  \n",
    "                    right_pos = int(hgvs_p.split(\"_\")[1].rstrip(\"del\")[2:])\n",
    "                    ref_peptide = ref_seq[(left_pos - self.flank_length):(right_pos + self.flank_length)]\n",
    "                    var_peptide =  var_seq[(left_pos - self.flank_length):(left_pos + self.flank_length)]\n",
    "\n",
    "            if (ref_peptide != \"\") and (var_peptide != \"\"):\n",
    "                non_standard_hgvs_p = convert_hgvs_to_non_standard(hgvs_p)\n",
    "                ref_peptides.append(f'>Ref {transcript_id}\\n{ref_peptide}')\n",
    "                var_peptides.append(f'>Var {transcript_id}({gene_name}):{hgvs_p} {non_standard_hgvs_p} \\n{var_peptide}')\n",
    "            else:\n",
    "                print(f'{hgvs_p} was filtered, please check it' )\n",
    "        return ref_peptides, var_peptides\n",
    "\n",
    "    def write_fasta(self, ref_peptides, var_peptides):\n",
    "        with open(f'{self.out_path}/{self.prefix}.ref_variants.faa', 'w') as outfile:\n",
    "            for ref_peptide, var_peptide  in zip(ref_peptides, var_peptides):\n",
    "                outfile.write(ref_peptide + \"\\n\")\n",
    "                outfile.write(var_peptide + \"\\n\")\n",
    "\n",
    "    def run(self):\n",
    "        self.read_csv()\n",
    "        \n",
    "        self.read_fasta()\n",
    "        ref_peptides, var_peptides = self.get_var_peptides()\n",
    "        ref_peptides, var_peptides = self.dedup_var_peptides(ref_peptides, var_peptides)\n",
    "        self.write_fasta(ref_peptides, var_peptides)\n",
    "        print(f'已将{len(ref_peptides)}条Ref多肽段写入{self.out_path}/{self.prefix}.ref.fasta文件。')\n",
    "        print(f'已将{len(var_peptides)}条Variant多肽段写入{self.out_path}{self.prefix}.variants.fasta文件。')\n",
    "if __name__ == '__main__':\n",
    "    generator = MutationPeptideGenerator('mutations.csv', 'input.fasta', 'output_dir',\"prefix\")\n",
    "    #generator.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2fb52c4d-4f2f-4f2d-be9f-670e35b409a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_005663.5_p.Ser379Ser not in fasta_seqs\n",
      "p.Ser379Ser was filtered, please check it\n",
      "p.Tyr122* was filtered, please check it\n",
      "p.Tyr107* was filtered, please check it\n",
      "p.Tyr109* was filtered, please check it\n",
      "p.Tyr108* was filtered, please check it\n",
      "p.Tyr87* was filtered, please check it\n",
      "NM_001366652.1_p.Ala719Ala not in fasta_seqs\n",
      "p.Ala719Ala was filtered, please check it\n",
      "NM_001366655.1_p.Ala719Ala not in fasta_seqs\n",
      "p.Ala719Ala was filtered, please check it\n",
      "NM_001366657.1_p.Ala707Ala not in fasta_seqs\n",
      "p.Ala707Ala was filtered, please check it\n",
      "NM_025153.3_p.Ala719Ala not in fasta_seqs\n",
      "p.Ala719Ala was filtered, please check it\n",
      "NM_001321866.2_p.Ile364Ile not in fasta_seqs\n",
      "p.Ile364Ile was filtered, please check it\n",
      "NM_001321867.2_p.Ile364Ile not in fasta_seqs\n",
      "p.Ile364Ile was filtered, please check it\n",
      "NM_198457.4_p.Ile295Ile not in fasta_seqs\n",
      "p.Ile295Ile was filtered, please check it\n",
      "已将13条Ref多肽段写入/home/jovyan/work/10.data_CODA_ahslyy/03.Result.SAKit2/0014274065_CA/results/08.SvInDelSnvCalling//COAD_0014274065.ref.fasta文件。\n",
      "已将13条Variant多肽段写入/home/jovyan/work/10.data_CODA_ahslyy/03.Result.SAKit2/0014274065_CA/results/08.SvInDelSnvCalling/COAD_0014274065.variants.fasta文件。\n"
     ]
    }
   ],
   "source": [
    "samplename = \"0014274065\"\n",
    "muation_csv = f\"/home/jovyan/work/10.data_CODA_ahslyy/03.Result.SAKit2/{samplename}_CA/results/08.SvInDelSnvCalling/COAD_{samplename}_somatic_filtered.csv\"\n",
    "aa_fasta = f\"/home/jovyan/work/10.data_CODA_ahslyy/05.SNVINDEL/COAD_{samplename}.single_variant.prot.fa\"\n",
    "outpath = f\"/home/jovyan/work/10.data_CODA_ahslyy/03.Result.SAKit2/{samplename}_CA/results/08.SvInDelSnvCalling/\"\n",
    "prefix = f\"COAD_{samplename}\"\n",
    "MutationPeptideGenerator(muation_csv, aa_fasta, outpath, prefix, encoding=\"gbk\").run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abedde67-a8eb-46b3-8aac-3d93b418c4a9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
